SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc

```diff
@@ -0,0 +1,453 @@
+  edm::LogInfo("CaloTruthAccumulator") << "Adding " << output_.pSimClusters->size() 
+				       << " SimParticles and " << output_.pCaloParticles->size()
+				       << " CaloParticles to the event.";
+
+  // now we need to normalize the hits and energies into hits and fractions
+  // (since we have looped over all pileup events)
+
+  for( auto& sc : *(output_.pSimClusters) ) {
+    auto hitsAndEnergies = sc.hits_and_fractions();
+    sc.clearHitsAndFractions();
+    for( auto& hAndE : hitsAndEnergies ) {
+      const float fraction = hAndE.second/m_detIdToTotalSimEnergy[hAndE.first];
+      sc.addRecHitAndFraction(hAndE.first,fraction);
+    }
+  }
+  
+  // save the SimCluster orphan handle so we can fill the calo particles
+  auto scHandle = event.put( std::move(output_.pSimClusters), "MergedCaloTruth" );
+  
+  // now fill the calo particles
+  for( unsigned i = 0; i < output_.pCaloParticles->size(); ++i ) {
+    auto& cp = (*output_.pCaloParticles)[i];
+    for( unsigned j = m_caloParticles.sc_start_[i]; j < m_caloParticles.sc_stop_[i]; ++j ) {
+      edm::Ref<SimClusterCollection> ref(scHandle,j);
+      cp.addSimCluster(ref);
+    }
+  }  
+
+  event.put( std::move(output_.pCaloParticles), "MergedCaloTruth" );
+
+  calo_particles().swap(m_caloParticles);
+
+  std::unordered_map<Index_t,float>().swap(m_detIdToTotalSimEnergy);
+
+  std::unordered_map<Barcode_t,Index_t>().swap(m_genParticleBarcodeToIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_simTrackBarcodeToIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_genBarcodeToSimTrackIndex);
+  std::unordered_map<Barcode_t,Index_t>().swap(m_simVertexBarcodeToIndex);
+
+  std::unordered_multimap<Index_t,Index_t>().swap(m_detIdToCluster);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simHitBarcodeToIndex);
+  std::unordered_multimap<Barcode_t,Barcode_t>().swap(m_simVertexBarcodeToSimTrackBarcode);
+  std::unordered_map<Barcode_t,Barcode_t>().swap(m_simTrackBarcodeToSimVertexParentBarcode);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simTrackToSimVertex);
+  std::unordered_multimap<Barcode_t,Index_t>().swap(m_simVertexToSimTrackParent); 
+  std::vector<Barcode_t>().swap(m_simVertexBarcodes);
+  std::unordered_map<Index_t,float>().swap(m_detIdToTotalSimEnergy);
+}
+
+template<class T> 
+void CaloTruthAccumulator::accumulateEvent( const T& event, 
+					    const edm::EventSetup& setup, 
+					    const edm::Handle< edm::HepMCProduct >& hepMCproduct) {
+  //
+  // Get the collections
+  //
+  
+  edm::Handle< std::vector<reco::GenParticle> > hGenParticles;
+  edm::Handle< std::vector<int> > hGenParticleIndices;
+  
+  event.getByLabel( simTrackLabel_, hSimTracks );
+  event.getByLabel( simVertexLabel_, hSimVertices );
+  
+  try {
+    event.getByLabel( genParticleLabel_, hGenParticles );
+    event.getByLabel( genParticleLabel_, hGenParticleIndices );
+  } catch( cms::Exception& exception ) {
```

Comment:
Send me your cmsDriver command and I'll check it out. I didn't have a
problem.

(Sent from my Nexus 6)

On Aug 7, 2016 10:30 AM, "Slava Krutelyov" <notifications@github.com> wrote:

> In SimGeneral/CaloAnalysis/plugins/CaloTruthAccumulator.cc
> <https://github.com/cms-sw/cmssw/pull/15293#discussion_r73803964>:
>
> > +					    const edm::EventSetup& setup,
> > +					    const edm::Handle< edm::HepMCProduct >& hepMCproduct) {
> > +  //
> > +  // Get the collections
> > +  //
> > +
> > +  edm::Handle< std::vector<reco::GenParticle> > hGenParticles;
> > +  edm::Handle< std::vector<int> > hGenParticleIndices;
> > +
> > +  event.getByLabel( simTrackLabel_, hSimTracks );
> > +  event.getByLabel( simVertexLabel_, hSimVertices );
> > +
> > +  try {
> > +    event.getByLabel( genParticleLabel_, hGenParticles );
> > +    event.getByLabel( genParticleLabel_, hGenParticleIndices );
> > +  } catch( cms::Exception& exception ) {
>
> it looks like I'm getting an uncaught exception here while testing on
> pileup setup
>
> *** THROW by cmsRun [thread 139864339315648 pid 36321]:
>  Exception of type N3cms9ExceptionE (address 0x7f3417ae0e00)
>  Destructor cms::Exception::~Exception() (0x7f34b57ebb70)
>  Stack:
>     1: 0x7f34989b6f13 void CaloTruthAccumulator::accumulateEvent<PileUpEventPrincipal>(PileUpEventPrincipal const&, edm::EventSetup const&, edm::Handle<edm::HepMCProduct> const&) + 931 [XX/lib/slc6_amd64_gcc530/pluginSimGeneralCaloAnalysisPlugins.so + 114451]
>     2: 0x7f34989acb36 CaloTruthAccumulator::accumulate(PileUpEventPrincipal const&, edm::EventSetup const&, edm::StreamID const&) + 358 [XX/lib/slc6_amd64_gcc530/pluginSimGeneralCaloAnalysisPlugins.so + 72502]
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/15293/files/838ad774de1cd5bf024063225a320c946e41a95b#r73803964>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBMOSBOmC5PRVpJ8h8tHX7M1iDCb0Fnks5qdfoZgaJpZM4JVi3o>
> .
>